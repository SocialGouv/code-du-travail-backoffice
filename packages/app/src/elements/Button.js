import styled from "@emotion/styled";
import PropTypes from "prop-types";
import React from "react";
import { Button as _Button } from "rebass";

import Icon from "./Icon";

const Container = styled(_Button)`
  background-color: ${p => p.backgroundColor};
  border-radius: 0.125rem;
  color: ${p => p.foregroundColor};
  cursor: ${p => (p.isDisabled ? "not-allowed" : "pointer")};
  display: flex;
  font-size: 0.9375rem;
  font-weight: 600;
  margin-left: ${p => (p.withLeftMargin ? "1rem" : 0)};
  margin-right: ${p => (p.withRightMargin ? "1rem" : 0)};
  min-height: 1.25rem;
  min-width: 1.25rem;
  opacity: ${p => (p.isDisabled ? 0.25 : 0.75)};
  padding: ${p => (p.hasText ? (p.isSmall ? "0.1rem 0.4rem 0.15rem" : "0.3rem 1rem 0.3rem") : "0")};
  white-space: nowrap;

  :hover {
    opacity: ${p => (p.isDisabled ? 0.25 : 1)};
  }

  svg {
    cursor: ${p => (p.isDisabled ? "not-allowed" : "pointer")};
    height: 1rem;
    margin-right: ${p => (p.hasText ? "0.25rem" : 0)};
    padding-top: ${p => (p.hasText ? "0.125rem" : 0)};
    width: 1rem;
  }
`;

const COLOR = {
  danger: { isLight: false, value: "var(--color-shadow)" },
  info: { isLight: false, value: "var(--color-misty-moss)" },
  primary: { isLight: false, value: "var(--color-lapis-lazuli)" },
  secondary: { isLight: true, value: "var(--color-periwinkle)" },
  warning: { isLight: false, value: "var(--color-lapis-lazuli)" },
};

const Button = ({
  children,
  color = "primary",
  icon,
  isDisabled = false,
  isSmall = false,
  isTransparent = false,
  onClick = () => undefined,
  withLeftMargin = false,
  withRightMargin = false,
  ...props
}) => {
  const hasText = children !== undefined;
  const isLight = COLOR[color].isLight;
  const backgroundColor = isTransparent ? "transparent" : COLOR[color].value;
  const foregroundColor = isTransparent
    ? COLOR[color].value
    : isLight
    ? "var(--color-eerie-black)"
    : "white";

  return (
    <Container
      backgroundColor={backgroundColor}
      disabled={isDisabled}
      foregroundColor={foregroundColor}
      hasText={hasText}
      isDisabled={isDisabled}
      isSmall={isSmall}
      onClick={isDisabled ? () => undefined : onClick}
      withLeftMargin={withLeftMargin}
      withRightMargin={withRightMargin}
      {...props}
    >
      {icon !== undefined && <Icon color={foregroundColor} icon={icon} />}
      {children}
    </Container>
  );
};

Button.propTypes = {
  color: PropTypes.string,
  icon: PropTypes.string,
  isDisabled: PropTypes.bool,
  isSmall: PropTypes.bool,
  isTransparent: PropTypes.bool,
  withLeftMargin: PropTypes.bool,
  withRightMargin: PropTypes.bool,
};

export default Button;
