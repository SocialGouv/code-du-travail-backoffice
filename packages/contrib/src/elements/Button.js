import React from "react";
import { Button as _Button } from "rebass";
import styled from "styled-components";

import Icon from "./Icon";

const Button = styled(_Button)`
  background-color: ${({ color, hasText }) =>
    !hasText ? "transparent" : color};
  border-radius: 0.125rem;
  color: ${({ isLight }) => (isLight ? "var(--color-eerie-black)" : "white")};
  cursor: ${({ disabled }) => (disabled ? "not-allowed" : "pointer")};
  display: flex;
  font-size: 0.9375rem;
  font-weight: 600;
  margin-right: ${({ hasGroup }) => (hasGroup ? "1rem" : 0)};
  min-height: 1.25rem;
  min-width: 1.25rem;
  opacity: ${({ disabled }) => (disabled ? 0.25 : 1)};
  padding: ${({ isSmall }) =>
    isSmall ? "0.1rem 0.5rem 0.15rem" : "0.3rem 1rem 0.3rem"};
  white-space: nowrap;

  :hover {
    opacity: ${({ disabled }) => (disabled ? 0.25 : 1)};
  }

  svg {
    cursor: ${({ disabled }) => (disabled ? "not-allowed" : "pointer")};
    margin-right: ${({ hasText }) => (hasText ? "0.25rem" : 0)};
    padding-top: ${({ hasText }) => (hasText ? "0.125rem" : 0)};
  }
`;

const COLOR = {
  danger: { isLight: false, value: "var(--color-shadow)" },
  info: { isLight: false, value: "var(--color-misty-moss)" },
  primary: { isLight: false, value: "var(--color-lapis-lazuli)" },
  secondary: { isLight: true, value: "var(--color-periwinkle)" },
  warning: { isLight: false, value: "var(--color-lapis-lazuli)" }
};

export default ({
  children,
  color = "primary",
  disabled = false,
  hasGroup = false,
  icon,
  isSmall = false,
  ...props
}) => {
  const isLight = COLOR[color].isLight;

  if (icon === undefined)
    return (
      <Button
        color={COLOR[color].value}
        disabled={disabled}
        hasGroup={hasGroup}
        hasText
        isLight={isLight}
        isSmall={isSmall}
        {...props}
      >
        {children}
      </Button>
    );

  const hasText = children !== undefined;

  if (hasText) {
    return (
      <Button
        color={COLOR[color].value}
        disabled={disabled}
        hasGroup={hasGroup}
        hasText
        isLight={isLight}
        isSmall={isSmall}
        {...props}
      >
        <Icon
          icon={icon}
          color={isLight ? "var(--color-eerie-black)" : "white"}
        />
        {children}
      </Button>
    );
  }

  return (
    <Button
      disabled={disabled}
      hasGroup={hasGroup}
      hasText={false}
      isLight={isLight}
      isSmall={isSmall}
      {...props}
    >
      <Icon icon={icon} color={COLOR[color].value} />
      {children}
    </Button>
  );
};
