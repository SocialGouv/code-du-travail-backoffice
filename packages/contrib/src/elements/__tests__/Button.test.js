import React from "react";
import { render } from "react-testing-library";
import Button from "../Button";

describe("[Contrib] elements/<Button />", () => {
  it("should match snapshot", () => {
    const { container } = render(<Button />);

    expect(container).toMatchSnapshot();
  });

  it("should match snapshot with a custom `color` prop", () => {
    const { container } = render(<Button color="secondary" />);

    expect(container).toMatchSnapshot();
  });

  it("should match snapshot with `hasGroup` prop", () => {
    const { container } = render(<Button hasGroup />);

    expect(container).toMatchSnapshot();
  });

  it("should match snapshot with a set `icon` prop", () => {
    const { container } = render(<Button icon="sync" />);

    expect(container).toMatchSnapshot();
  });

  it("should match snapshot with `disabled` prop", () => {
    const { container } = render(<Button disabled />);

    expect(container).toMatchSnapshot();
  });
});
