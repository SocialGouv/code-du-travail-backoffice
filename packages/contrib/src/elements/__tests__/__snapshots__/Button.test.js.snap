// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Contrib] elements/<Button /> should match snapshot 1`] = `
<div>
  <button
    class="sc-bdVaJa sc-EHOje Button-sc-1bpx84u-0 iDhkoI"
    color="var(--color-lapis-lazuli)"
    font-size="inherit"
    font-weight="bold"
  />
</div>
`;

exports[`[Contrib] elements/<Button /> should match snapshot 2`] = `
<div>
  <button
    class="sc-bdVaJa sc-EHOje Button-sc-1bpx84u-0 bSSmqr"
    color="var(--color-periwinkle)"
    font-size="inherit"
    font-weight="bold"
  />
</div>
`;

exports[`[Contrib] elements/<Button /> should match snapshot 3`] = `
<div>
  <button
    class="sc-bdVaJa sc-EHOje Button-sc-1bpx84u-0 kTjDPw"
    color="var(--color-lapis-lazuli)"
    font-size="inherit"
    font-weight="bold"
  />
</div>
`;
